src/QMCDrivers/DMC/WalkerControlMPI.cpp

```diff
@@ -73,63 +61,59 @@ WalkerControlMPI::WalkerControlMPI(Communicate* c): WalkerControlBase(c)
   setup_timers(myTimers, DMCMPITimerNames, timer_level_medium);
 }
 
-int
-WalkerControlMPI::branch(int iter, MCWalkerConfiguration& W, RealType trigger)
+/** perform branch and swap walkers as required
+ *
+ *  it takes 4 steps:
+ *    1. shortWalkers marks good and bad walkers.
+ *    2. allreduce and make the decision of load balancing.
+ *    3. send/recv walkers. Receiving side recycle bad walkers' memory first.
+ *    4. copyWalkers generate walker copies of good walkers.
+ */
+int WalkerControlMPI::branch(int iter, MCWalkerConfiguration& W, RealType trigger)
 {
-  DMC_BRANCH_START(Timer localTimer);
-  TinyVector<RealType,3> bTime(0.0);
   myTimers[DMC_MPI_branch]->start();
   myTimers[DMC_MPI_prebalance]->start();
   std::fill(curData.begin(),curData.end(),0);
-  //std::fill(NumPerNode.begin(),NumPerNode.end(),0);
   sortWalkers(W);
   //use NumWalkersSent from the previous exchange
   curData[SENTWALKERS_INDEX]=NumWalkersSent;
   //update the number of walkers for this node
   curData[LE_MAX+MyContext]=NumWalkers;
-  DMC_BRANCH_STOP(bTime[0],localTimer.elapsed());
-  DMC_BRANCH_START(localTimer.restart());
-  int nw = copyWalkers(W);
+  //myTimers[DMC_MPI_imbalance]->start();
+  //myComm->barrier();
+  //myTimers[DMC_MPI_imbalance]->stop();
+  myTimers[DMC_MPI_allreduce]->start();
   myComm->allreduce(curData);
+  myTimers[DMC_MPI_allreduce]->stop();
   measureProperties(iter);
   W.EnsembleProperty=EnsembleProperty;
-  DMC_BRANCH_STOP(bTime[1],localTimer.elapsed());
-  DMC_BRANCH_START(localTimer.restart());
-  ////update the samples and weights
-  //W.EnsembleProperty.NumSamples=curData[WALKERSIZE_INDEX];
-  //W.EnsembleProperty.Weight=curData[WEIGHT_INDEX];
-  //RealType wgtInv(1.0/curData[WEIGHT_INDEX]);
-  //accumData[ENERGY_INDEX]     += curData[ENERGY_INDEX]*wgtInv;
-  //accumData[ENERGY_SQ_INDEX]  += curData[ENERGY_SQ_INDEX]*wgtInv;
-  //accumData[WALKERSIZE_INDEX] += curData[WALKERSIZE_INDEX];
-  //accumData[WEIGHT_INDEX]     += curData[WEIGHT_INDEX];
   Cur_pop=0;
   for(int i=0, j=LE_MAX; i<NumContexts; i++,j++)
   {
     Cur_pop+= NumPerNode[i]=static_cast<int>(curData[j]);
   }
   myTimers[DMC_MPI_prebalance]->stop();
-  myTimers[DMC_MPI_loadbalance]->start();
   if(qmc_common.async_swap)
+  {
+    myTimers[DMC_MPI_copyWalkers]->start();
+    copyWalkers(W);
+    //myComm->barrier();
```

Comment:
What are these barriers for?

Reply:
When I investigate the source of imbalance, I turn on these barrier one at a time.

Reply:
Any need to keep them? We are trying to minimize / remove sections of commented out code.

Reply:
Not really. I will clean them up in the next PR which also touches this part. If I remove it here, it will be more complicated for me to do the rebase.